Jumpscale/clients/blockchain/tfchain/TFChainCapacity.py

```diff
@@ -63,7 +63,75 @@ def _grid_broker_pub_key(self):
             self._grid_broker_pub_key_ = vk.to_curve25519_public_key()
         return self._grid_broker_pub_key_
 
-    def reserve_s3(self, email, threebot_id, location, size=1, source=None, refund=None):
+    def _notary_data_get(self, hash_key):
+        data = self._notary_client.get(hash_key)
+        sk = self._threebot_singing_key(data.threebot_id)
+        pk = _signing_key_to_private_key(sk)
+        box = nacl.public.Box(pk, self._grid_broker_pub_key)
+        data_dict = j.data.serializers.msgpack.loads(box.decrypt(data.content))
+
+        return data.threebot_id, data_dict
+
+    def reservations_transactions_list(self):
+        """
+        list all transactions that happened due to a reservation
+        """
+        transactions = list()
+        for transaction in self._wallet.transactions:
+            if not transaction.data:
+                continue
+            try:
+                _, data_dict = self._notary_data_get(transaction.data.value.decode())
+                if data_dict.get("type") in ["vm", "reverse_proxy", "namespace", "s3"]:
+                    transactions.append(transaction)
+            except Exception as e:
+                if "reservation not found" not in str(e):
+                    raise e
+        return transactions
+
+    def reservation_extend(self, transaction, duration=1, source=None, refund=None):
```

Comment:
This function also needs to check that the reservation is not already expired. We cannot extend a expired transaction because we cannot ensure the capacity has not been already sell to someone else in the mean time.

Which also mean we need to be a bit smart in how we compute the final expiration date because if someone extend a 1 month reservation for another month, but we are only 2 week in the first reservation, the final expiration data is in 1month and 2 weeks.

Reply:
As discussed, I added this logic in the grid_broker template.